### PR TITLE
Disallow joining ended rooms in realtime multiplayer

### PR DESCRIPTION
**Pull request by frenzibyte**

### CODE REVIEW

File: osu.Game/Screens/Multi/Lounge/Components/RoomsContainer.cs [hunk 1]
```diff
@@ -115,7 +115,9 @@ private void addRooms(IEnumerable<Room> rooms)
                     {
                         if (room == selectedRoom.Value)
                         {
-                            joinSelected();
+                            if (roomManager.CanJoinRoom(room))
+                                joinSelected();
```
Comment (bdach) [inline osu.Game/Screens/Multi/Lounge/Components/RoomsContainer.cs:119]: I'm not convinced by the way this is done at all. This logic is all very far from `RoomManager`, wherein the request is being fired - it should be in `JoinRoom()` in my opinion.

There are a few ways to implement that, and I think that `CanJoinRoom` doesn't even maybe need to exist - can probably just override `RealtimeRoomManager`'s impl, like so:

```csharp
public override void JoinRoom(Room room, Action<Room> onSuccess = null, Action<string> onError = null)
{
    if (room.Status.Value is RoomStatusEnded)
    {
        onError?.Invoke("Cannot join room as it has already ended.");
        return;
    }

    base.JoinRoom(room, r => joinMultiplayerRoom(r, onSuccess, onError), onError);
}
```

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Multi/Lounge/Components/RoomsContainer.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Screens/Multi/Lounge/Components/RoomsContainer.cs:119]: Ah, was thinking `onError` is not the right way for this as this case is more of guaranteed while other cases of invoking should not be possible to happen, will change back...

**Comment by peppy** [comment]
I don't get it. What is this doing? The server already disallows this from happening right?

**Comment by frenzibyte** [comment]
Yeah but now it early checks for that and returns, without logging as an error and otherwise, thought this would be more correct than before now?

**Comment by peppy** [comment]
I don't see the point.

**Comment by frenzibyte** [comment]
Previously I made a method that checks logically whether it should be fine to join the room or not, to avoid making the join request at all otherwise, but https://github.com/ppy/osu/pull/11256#discussion_r547531356 tells that it's not really the way here.

**Comment by peppy** [comment]
Well it's a bit unfortunate how the flow of clicking rooms happens in the lounge right now (hard to handle in a more appropriate place). Maybe this is fine for now.

**Review by peppy — APPROVED** [review]

**Pull request merged**